### ISSUE TITLE
RetroArch freezes and does not return to the frontend

### ISSUE THREAD

**Issue by anderbubble** [eduardofilo/RG350_adam_image#3]
As indicated in the README:

> Sometimes, when trying to exit RetroArch, it freezes and does not return to the frontend. In these cases, force close the application in the foreground (RetroArch at that moment) by pressing the shortcut Power + Select.

Do you have any information on why this is happening? It happens pretty often for me. Maybe a little less than half of the times I try to quit RA. And I see it regardless of whether I try Start + Select or quit from the menu.

**Comment by eduardofilo** [comment]
No, I'm sorry.
The only clue that I have is that the hang is produced in the `exec.sh` script that is inside the OPK wrappers. This is before to exit game:

```
  370 od        1476 S<   {exec.sh} /bin/sh exec.sh /media/sdcard/roms/GB/Adventure Island (USA, Europe).zip
  371 od       35624 S<   /media/data/local/bin/retroarch_rg350_odbeta -v -L /media/data/local/home/.retroarch/cores/gambatte_libretro.so --config
```

and this is after:

```
  370 od        1476 S<   {exec.sh} /bin/sh exec.sh /media/sdcard/roms/GB/Adventure Island (USA, Europe).zip
  371 od           0 Z<   [retroarch_rg350]
```

I have been researching and it seems that sometimes the SIGCHLD signal is not received by my exec.sh script, and keep waiting, so it doesn't ends and SimpleMenu is not restored.

I have to research further. Any help will be welcome.

**Comment by anderbubble** [comment]
Can you point me to any documentation for how to get at this level of debug info? Is it just by ssh'ing in through the usb interface? (I haven't done that yet, but should be able to without much trouble.)

**Comment by eduardofilo** [comment]
Yes is a terminal session through SSH by USB cable. In Linux a terminal is essential.

**Comment by eduardofilo** [comment]
I comment on a finding made by Marlenus on Discord:

```
With ADAM I did find a lot of crashes when trying to exit games or apply video filters. These did not occur with OD Beta and Retroarch on its own, so I though perhaps Simplemenu caused this.

But when I deleted config folder and retroarch.cfg file as Eduardo showed above, all those crashes have gone away and it’s very stable now.

So maybe there is a retroarch setting ADAM uses that triggers some conflict on a game closedown?
```

**Comment by anderbubble** [comment]
I managed to get a shell on my 280v and experimented a bit.

It seems that I can cause a hang somewhat reliably by opening and closing the retroarch ui and immediately closing with start+select.

When I see a hang, I can clear the hang by issuing a `kill` to the `opkrun` process. It also appears that I gain a new overlapping `/opk` mount each time the hang happens.

After upgrading to the latest image I no longer see the `exec.sh` script. It looks like retroarch is run directly by the `.desktop` file. @eduardofilo, is that correct?

I also tried using `strace` to monitor the `{opkrun}` process. The process is running `wait4(` while retroarch is running, as expected. When retroarch exits successfully, `wait4(` completes as expected, and then a SIGCHLD is logged. But when the hang occurs, no SIGCHLD is logged and `wait4(` never completes.

I thought maybe the `{opkrun}` process was multi-threaded, and the wrong thread was getting the signal (https://unix.stackexchange.com/questions/225687/what-happens-to-a-multithreaded-linux-process-if-it-gets-a-signal) but that does not appear to be the case. (It's single threaded, according to `/proc/<pid>/tasks/`.)

@eduardofilo I think I've seen this written down somewhere, but now I can't find it: how does the start+select shortcut work? I'm not convinced that the shortcutting is the source of the problem, since I've had hangs by simply quitting RA from the menu. But it might help me track it down.

**Comment by anderbubble** [comment]
I think I've discovered something useful. `{opkrun}` is a single-threaded process; but the RA process itself is multi-threaded. It's got five threads for me under normal operation (when running the GBA core, anyway). When the hang occurs, the primary thread goes to Z state, but two more threads remain, in S state.

The first is waiting on a `futex(<address>, FUTEX_WAIT_PRIVATE, 7, NULL`.

The second is waiting on a `futex(<address>, FUTEX_WAIT_PRIVATE, 515, NULL`

I think that the primary thread can't return a code to its parent until all of its threads have terminated, and these two are locked.

This looks to me like a threading but in RA; but given the comments at https://github.com/eduardofilo/RG350_adam_image/issues/3#issuecomment-913154649 (and the fact that I have not observed this problem outside of the ADAM image) I expect that there is a RA setting being configured by this image that is triggering a RA threading bug.

@eduardofilo I tried deleting (actually, moving out-of-the-way) the `.retroarch` directory, hoping that it would generate a new default `retroarch.cfg` file and I could compare the two. But when I do that, a new `.retroarch` directory is populated but no new `retroarch.cfg` file appears and RA fails to start. How can I get a default retroarch.cfg file without your modifications to compare?

**Comment by anderbubble** [comment]
@eduardofilo I think the issue is the openal audio driver.

```
-audio_driver = "alsa"
+audio_driver = "openal"
```

I managed to get RA to generate a default config, and I diff'd the two. That's the main different that stuck out to me.

I reverted to `audio_driver=alsa` and now I can't get it to hang.

**Comment by eduardofilo** [comment]
Sorry for the late reply @anderbubble. I answer your questions.
Yes, the exec.sh script was removed in an attempt to fix this issue.
The Select+Start shortcut is configured inside RetroArch. As you say I never believed that was the problem.
The resfreshed cfg file can be obtained simply deleting it, not the entire directory. One hour ago, I have done precisely that and I have seen that the problem disappears (apparently). Using diff between new and old files It has caught my attention the audio_driver difference also. I wish I had read your comments earlier.

Naturally the next release will bring this changes. Thanks for your effort in solving this. I attach the file that works to me in the meantime.

[retroarch.cfg.zip](https://github.com/eduardofilo/RG350_adam_image/files/7129310/retroarch.cfg.zip)

**Comment by eduardofilo** [comment]
Fixed (I hope) in v1.2

**Comment by RobinBot89** [comment]
I had this issue too on adam 1.0.3. I just upgraded to 1.3 and its fixed.
